test/scripts/e2e_subs/app-inner-calls-csp.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+
+import os
+import sys
+from goal import Goal
+import algosdk.logic as logic
+
+from datetime import datetime
+
+stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+print(f"{os.path.basename(sys.argv[0])} start {stamp}")
+
+goal = Goal(sys.argv[1], autosend=True)
+
+joe = goal.new_account()
+
+txinfo1, err = goal.pay(goal.account, joe, amt=500_000)
+assert not err, err
+
+txinfo1, err = goal.keyreg(joe, nonpart=True)
+assert not err, err
+joeb = goal.balance(joe)
+
+# app1 calls the clear state program of app2 (after opting into it) which issues an
+# inner app call to app3. This verifies that both accessing a CSP through inner app
+# calls and issuing inner app calls from a CSP is possible.
+app1 = """
+#pragma version 6
+ txn ApplicationID
+ bz end
+
+ itxn_begin
+  int appl
+  itxn_field TypeEnum
+
+  txn Applications 1
+  dup
+  store 0
+  itxn_field ApplicationID
+
+  int OptIn
+  itxn_field OnCompletion
+
+  txn Applications 2
+  itxn_field Applications
+
+ itxn_next
+  int appl
+  itxn_field TypeEnum
+
+  load 0
+  itxn_field ApplicationID
+
+  txn Applications 2
+  itxn_field Applications
+
+  int ClearState
+  itxn_field OnCompletion
+
+ itxn_submit
+
+
+end:
+ int 1
+"""
+
+app2 = """
+#pragma version 6
+ txn ApplicationID
+ bz end
+
+ itxn_begin
+  int appl
+  itxn_field TypeEnum
+
+  txn Applications 1
+  itxn_field ApplicationID
+ itxn_submit
+
+
+ end:
+  int 1
+"""
+
+app3 = """
+#pragma version 6
+int 1
+"""
+
+goal.autosend = True
+
+txinfo1, err = goal.app_create(joe, goal.assemble(app1))
+assert not err, err
+app1ID = txinfo1['application-index']
+assert app1ID
+
+# insert clear state program with inner app call
+txinfo2, err = goal.app_create(joe, goal.assemble(app2), goal.assemble(app2))
+assert not err, err
+app2ID = txinfo2['application-index']
+assert app2ID
+
+# dummy destination app
+txinfo3, err = goal.app_create(joe, goal.assemble(app3))
+assert not err, err
+app3ID = txinfo3['application-index']
+assert app3ID
+
+# fund the apps
+txinfo1, err = goal.pay(goal.account, goal.app_address(app1ID), amt=4_000_000)
+assert not err, err
+
+txinfo2, err = goal.pay(goal.account, goal.app_address(app2ID), amt=4_000_000)
+assert not err, err
+
+# execute c2c w/ CSP
+start_app, err = goal.app_call(joe, app1ID, foreign_apps=[int(app2ID), int(app3ID)])
+assert not err, err
+
```

Comment:
@algoidurovic I'm wondering if you considered an assertion to (further) prove _app3_ is invoked by CSP successfully?

* I'm interpreting that _no_ error on line 118 implies successful inner-transaction CSP invocation.
* I know you mentioned there's ambiguity when deciding scope of an E2E test.  
* I'm also unclear on how far we'd expect to go.  Prior to seeing the assertions, I imagined asserting against modified _app2_ or _app3_ state.  Though I'm open to being convinced the additional assertion is extraneous.

Reply:
Right, no error implies the entire chain of app calls executed successfully, including the call to app2's CSP and the call from the CSP to app3. The only app that has any state change should be app2 because its CSP is invoked leading to app1's OptIn status with app2 being revoked. This can be enforced by calling app2's CSP from app1 again and asserting that an error occurred. Is this what you had in mind?

Reply:
@algoidurovic Gotcha - Below I share what I'm thinking with a  disclaimer:  
* I don't feel experienced enough to know if the proposed assertions are _reasonable_ or considered out-of-scope/duplicative.
* My starting assumption is:  The proposed assertions are _reasonable_ though I'll follow guidance from more experienced team members.

Here's what I had in mind.  If easier to discuss live, let me know.
* I'm imagining adding a side effect to _app3_ that can be observed after executing the app call. 
* An example I've seen comes from @ahangsu's recent tests:  https://github.com/ahangsu/c2c-testscript/blob/f3d5f9e33cfccc280c96b481be58f2cef3df0148/c2c_test/inner_256/app_test.py#L97-L102.  The idea applied here is to store an invocation counter in global state that can be read by the test.

Reply:
@algoidurovic Re-opening because I still don't believe the test confirms the following case:  _Issuing inner app calls from a CSP is possible_.

Let me know what you think.  I can discuss live if the thinking doesn't make sense.

Rationale:
Since the CSP is intended to _always_ complete successfully, I think the test ought to assert a side-effect of calling _app3_ happens.  Side-effects include a global invocation counter as mentioned in the prior comment.

Reply:
@algoidurovic Thanks for the changes in https://github.com/algorand/go-algorand/pull/3693/commits/287918887f2dc85015ce2893ed0e00e1d2e6a1f8.  The commit adds the kind of testing I had in mind  - Resolving.